### PR TITLE
[X86][CodeGen] Support using NF instructions for flag copy lowering

### PR DESCRIPTION
**Pull request by KanRobert**

### CODE REVIEW

**Comment by KanRobert** [comment]
@phoebewang @e-kud @FreddyLeaf This PR is ready for review.

File: llvm/test/CodeGen/X86/apx/flags-copy-lowering.ll [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+; CHECK-NEXT:    movq %rcx, 16(%rax)
+; CHECK-NEXT:    movq %rsi, (%rax)
+; CHECK-NEXT:    movq %rdi, 24(%rax)
+; CHECK-NEXT:    movq %r8, 8(%rax)
+; CHECK-NEXT:    retq
+  %z = call <2 x i128> @llvm.ssub.sat.v2i128(<2 x i128> %x, <2 x i128> %y)
+  ret <2 x i128> %z
+}
+
+; TODO: Remove the 2nd cmpl by using NF imul.
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/apx/flags-copy-lowering.ll:53]: I checked with this, firstly this was done by `sinkCmpExpression` in CodeGenPrepare.
But even we don't do the sink, e.g., https://godbolt.org/z/fMbsnrYdo
We still cannot pass it through `EFLAGS`.

That means, We can always assume the def and use of `EFLAGS` occur within the same BB. So you don't need to traverse BBs to search the dead def EFLAGS. Searching inside the BB is enough. Or if you have concern, you can add an assert to make sure of def/use of `EFLAGS` are in the same BB.

File: llvm/lib/Target/X86/X86FlagsCopyLowering.cpp [hunk 2]
```diff
@@ -254,14 +268,102 @@ bool X86FlagsCopyLoweringPass::runOnMachineFunction(MachineFunction &MF) {
...
+             I != E; ++I) {
+          MachineInstr &MI = *I;
+          EFLAGSClobber ClobberType = getClobberType(MI);
+          if (ClobberType == NoClobber)
+            continue;
+
+          if (ClobberType == InevitableClobber)
+            goto ProcessNextCopyI;
+
+          assert(ClobberType == EvitableClobber && "unexpected workflow");
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86FlagsCopyLowering.cpp:325]: Can't this be failed? I think we cannot transform another `$eflags = COPY %x` in `MBB` at least, right?

File: llvm/test/CodeGen/X86/apx/flags-copy-lowering.ll (hunk 1 above)
Comment (KanRobert) [reply llvm/test/CodeGen/X86/apx/flags-copy-lowering.ll:53]: Offline talked with Phoebe. We need to traverse BBs, and at the same time, we need to update liveins of BBs.

File: llvm/lib/Target/X86/X86FlagsCopyLowering.cpp (hunk 2 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/X86FlagsCopyLowering.cpp:325]: Then it's a InevitableClobber and handled at line 323.

File: llvm/test/CodeGen/X86/apx/flags-copy-lowering.ll (hunk 1 above)
Comment (KanRobert) [reply llvm/test/CodeGen/X86/apx/flags-copy-lowering.ll:53]: Done @phoebewang

**Comment by KanRobert** [comment]
Ping @phoebewang ?

**Review by phoebewang — APPROVED** [review]
LGTM.

**Pull request merged**